### PR TITLE
[SPARK-34002][SQL] Fix the usage of encoder in ScalaUDF

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch fixes few issues when using encoders to serialize input/output in `ScalaUDF`.

### Why are the changes needed?

This fixes a bug when using encoders in Scala UDF. First, the output data type should be corrected to the corresponding data type of the object serializer. Second, `catalystConverter` should not serialize `Option[_]` as the ordinary row because in `ScalaUDF` case it is serialized to a column, not the top-level row. Otherwise, there will be a redundant `value` struct wrapping the serialized `Option[_]` object.

### Does this PR introduce _any_ user-facing change?

Yes, fixing a bug of `ScalaUDF`.

### How was this patch tested?

Unit test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 1]
```diff
@@ -304,8 +304,13 @@ case class ExpressionEncoder[T](
...
 
+  /**
+   * This is used for `ScalaUDF` (see `UDFRegistration`). As the serialization in `ScalaUDF` is for
+   * individual column, not the whole row, we just take the data type of vanilla object serializer,
+   * not `serializer` which is transformed somehow.
+   */
   def dataTypeAndNullable: Schema = {
-    val dataType = if (isSerializedAsStruct) schema else schema.head.dataType
+    val dataType = objSerializer.dataType
     Schema(dataType, objSerializer.nullable)
```
[the remark this answers is not in the document]
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:314]: Btw, we don't need to move this thin helper func into the `UDFRegistration` side? It seems this func exists now for the specific use for udf-related code.

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 2]
```diff
@@ -1982,8 +1982,34 @@ class DatasetSuite extends QueryTest
...
+    def f1(bar: Bar): Option[Bar] = {
+      None
+    }
+
+    def f2(bar: Option[Bar]): Option[Bar] = {
+      bar
+    }
+
+    val udf1: UserDefinedFunction = udf(f1 _).withName("f1")
+    val udf2: UserDefinedFunction = udf(f2 _).withName("f2")
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:1996]: nit: `val udf1 = udf(f1 _).withName("f1")`

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 3]
```diff
@@ -1982,8 +1982,34 @@ class DatasetSuite extends QueryTest
...
+
+    val df = (1 to 2).map(i => Tuple1(Bar(1))).toDF("c0")
+    val newDf = df
+      .withColumn("c1", udf1(col("c0")))
+      .withColumn("c2", udf2(col("c1")))
+    val schema = newDf.schema
+    assert(schema == StructType(
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:2003]: `assert(newDf.schema == StructType(`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala:314]: Okay, looks putting it in `UDFRegistration` is better.

**Comment by HyukjinKwon** [comment]
cc @Ngone51 and @cloud-fan FYI.

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala [hunk 4]
```diff
@@ -1121,3 +1123,17 @@ class UDFRegistration private[sql] (functionRegistry: FunctionRegistry) extends
   // scalastyle:on line.size.limit
 
 }
+
+object UDFRegistration {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala:1127]: nit: 
```suggestion
private object UDFRegistration {
```

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala:1127]: nit:

```suggestion
private[sql] object UDFRegistration {
```

**Comment by viirya** [comment]
retest this please

**Review by srowen — COMMENTED** [review]
From the description, does that not change the type of the object returned from a UDF? just checking whether this introduces a breaking change.

**Comment by viirya** [comment]
> From the description, does that not change the type of the object returned from a UDF? just checking whether this introduces a breaking change.

No, this does not change the type of object returned. Actually we introduced a breaking change at #28979, which also causes this bug. That is not shipped out yet and only in 3.1 branch. This patch fixes it and keeps the type of object returned as Spark 3.0.

File: sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala [hunk 5]
```diff
@@ -1121,3 +1123,17 @@ class UDFRegistration private[sql] (functionRegistry: FunctionRegistry) extends
...
+
+private[sql] object UDFRegistration {
+  /**
+   * Obtaining the schema of output encoder for `ScalaUDF`.
+   *
+   * As the serialization in `ScalaUDF` is for individual column, not the whole row,
+   * we just take the data type of vanilla object serializer, not `serializer` which
+   * is transformed somehow for top-level row.
+   */
+  def outputSchema(outputEncoder: ExpressionEncoder[_]): ScalaReflection.Schema = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala:1135]: shall we define it in `ExpressionEncoder`?

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala:1135]: I moved it out per https://github.com/apache/spark/pull/31103#discussion_r554405600. Do you think it is better to put in `ExpressionEncoder`?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala:1135]: ah makes sense, it's udf registration specific. Let's keep it here.

**Review by cloud-fan — APPROVED** [review]
good catch!

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @viirya and all.

I marked SPARK-34002 as a blocker for Apache Spark 3.1.1 because this is a regression due to SPARK-32154 at 3.1.0 only.

**Comment by dongjoon-hyun** [comment]
Merged to master/3.1.

**Comment by viirya** [comment]
Thanks @dongjoon-hyun @cloud-fan @maropu @srowen @HyukjinKwon

**Comment by Ngone51** [comment]
Late LGTM. Thanks for the fix @viirya !

**Pull request closed without merging**